pallets/rolldown/src/lib.rs

```diff
@@ -0,0 +1,916 @@
+
+	#[pallet::storage]
+	#[pallet::unbounded]
+	#[pallet::getter(fn get_pending_updates_u256_array)]
+	pub type PendingUpdatesU256Array<T: Config> =
+		StorageMap<_, Blake2_128Concat, BlockNumberFor<T>, Vec<U256>, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::unbounded]
+	#[pallet::getter(fn get_hash_pending_updates_u256_array)]
+	pub type HashPendingUpdatesU256Array<T: Config> =
+		StorageMap<_, Blake2_128Concat, BlockNumberFor<T>, U256, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::unbounded]
+	#[pallet::getter(fn get_pending_requests)]
+	pub type PENDING_REQUESTS<T: Config> =
+		StorageMap<_, Blake2_128Concat, BlockNumberFor<T>, (T::AccountId, String), OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::unbounded]
+	#[pallet::getter(fn get_sequencer_rights)]
+	pub type SEQUENCER_RIGHTS<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, SequencerRights, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::unbounded]
+	#[pallet::getter(fn get_pending_updates)]
+	pub type PENDING_UPDATES<T: Config> =
+		StorageMap<_, Blake2_128Concat, u128, Update<T::AccountId>, OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {}
+
+	#[pallet::error]
+	/// Errors
+	pub enum Error<T> {
+		OperationFailed,
+	}
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		type AddressConverter: TryConvert<String, Self::AccountId>;
+		type SequencerStakingProvider: SequencerStakingProviderTrait<
+			Self::AccountId,
+			BalanceOf<Self>,
+		>;
+		// Dummy so that we can have the BalanceOf type here for the SequencerStakingProviderTrait
+		type Tokens: MultiTokenCurrency<Self::AccountId>
+			+ MultiTokenReservableCurrency<Self::AccountId>
+			+ MultiTokenCurrencyExtended<Self::AccountId>;
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::DbWeight::get().reads_writes(1, 1).saturating_add(Weight::from_parts(40_000_000, 0)))]
+		// sequencer read
+		//fn update_l2_from_l1(origin: String, l1_pending_requests_json: &'static str, current_block_number: u32) {
+		pub fn update_l2_from_l1(
+			origin: OriginFor<T>,
+			input_json: String,
+		) -> DispatchResultWithPostInfo {
+			let sequencer = ensure_signed(origin)?;
+			// check json length to prevent big data spam, maybe not necessary as it will be checked later and slashed
+			let current_block_number = <frame_system::Pallet<T>>::block_number();
+			let dispute_period_end = current_block_number + (DISPUTE_PERIOD_LENGTH as u32).into();
+
+			let l1_pending_requests_json = input_json.trim().to_string();
+
+			// ensure sequencer has rights to update
+			if let Some(sequencer) = SEQUENCER_RIGHTS::<T>::get(&sequencer) {
+				if sequencer.readRights == 0 {
+					log!(info, "{:?} does not have sufficient readRights", sequencer);
```

Comment:
could it happen that all the sequencers end up with zero reads ? what would happen if so?

Reply:
yes it could, there is an emergency root only fn, which will do council_update_l2_from_l1. Same as regular, but requests from this json are not going to the queue and cannot be canceled, they are executed right away. Amongst these request are multiple cancel resolutions (if there are no read rights, there must have been a lot of cancels). When cancel resolutions are processed, slashing occures and read rights are returned.